src/puppetlabs/puppetdb/queue.clj

```diff
@@ -0,0 +1,199 @@
+            [puppetlabs.puppetdb.nio :refer [get-path]]))
+
+(def metadata-command-names
+  (vals constants/command-names))
+
+(defn stream->json [^InputStream stream]
+  (try
+    (-> stream
+        (InputStreamReader. StandardCharsets/UTF_8)
+        BufferedReader.
+        (json/parse-stream true))
+    (catch Exception e
+      (throw+ {:kind ::parse-error} e "Error parsing command"))))
+
+(defn metadata-str [received command version certname]
+  (format "%d_%s_%d_%s.json"
+          (tcoerce/to-long received) command version certname))
+
+(defn- metadata-rx [valid-commands]
+  (re-pattern (str
+               "([0-9]+)_("
+               (str/join "|" (map #(format "(?:%s)" (re-quote-replacement %))
+                                  valid-commands))
+               ")_([0-9]+)_(.*)\\.json")))
+
+(defn metadata-parser
+  ([] (metadata-parser metadata-command-names))
+  ([valid-commands]
+   ;; NOTE: changes here may affect the DLO, e.g. it currently assumes
+   ;; the trailing .json.
+   (let [rx (metadata-rx valid-commands)]
+     (fn [s]
+       (when-let [[_ stamp command version certname] (re-matches rx s)]
+         (and certname
+              {:received (-> stamp
+                             Long/parseLong
+                             tcoerce/from-long
+                             kitchensink/timestamp)
+               :version (Long/parseLong version)
+               :command command
+               :certname certname}))))))
+
+(def parse-metadata (metadata-parser))
+
+(def parse-cmd-filename
+  (let [parse-metadata (metadata-parser (cons "unknown" metadata-command-names))]
+    (fn [s]
+      (let [rx #"([0-9]+)-(.*)"]
+        (when-let [[_ id qmeta] (re-matches rx s)]
+          (parse-metadata qmeta))))))
+
+(defrecord CommandRef [id command version certname received callback annotations delete?])
+
+(defn cmdref->entry [{:keys [id command version certname received]}]
+  (stock/entry id (metadata-str received command version certname)))
+
+(defn entry->cmdref [entry]
+  (let [{:keys [received command version certname]} (-> entry
+                                                        stock/entry-meta
+                                                        parse-metadata)]
+    (map->CommandRef {:id (stock/entry-id entry)
+                      :command command
+                      :version version
+                      :certname certname
+                      :received received
+                      :callback identity
+                      :annotations {:id (stock/entry-id entry)
+                                    :received received}})))
+
+(defn cmdref->cmd [q cmdref]
+  (let [entry (cmdref->entry cmdref)]
+    (with-open [command-stream (stock/stream q entry)]
+      (assoc cmdref
+             :payload (stream->json command-stream)
+             :entry entry))))
+
+(defn cons-attempt [cmdref exception]
+  (update cmdref :attempts conj {:exception exception
```

Comment:
We certainly could; whether that's a win or a loss I suppose depends on what fraction of commands have failures before being processed.